interpretation/train_classifier.py

```diff
@@ -0,0 +1,219 @@
+    train = pysero_df_pivot.iloc[train_ids]
+    test = pysero_df_pivot.iloc[test_ids]
+    #%% Initialize xgboost tree classifier
+    clf = XGBClassifier(
+     learning_rate=0.01,
+     n_estimators=1000,
+     max_depth=5,
+     min_child_weight=1,
+     gamma=0,
+     subsample=0.8,
+     colsample_bytree=0.8,
+     objective='binary:logistic',
+     nthread=8,
+     scale_pos_weight=1,
+     seed=0)
+    #%% set up CV folds by serum ID
+    gkf = GroupKFold(n_splits=4)
+    folds = []
+    for fold in gkf.split(train, groups=train['serum ID']):
+        folds.append(fold)
+    #%% Tune xgboost parameter
+    param_tests = [{
+     'max_depth': range(1, 8, 1),
+     'min_child_weight':range(1, 8, 1)
+    },
+    {
+     'gamma':[i/10.0 for i in range(0,5)]
+    },
+    {
+     'subsample':[i/10.0 for i in range(6,10)],
+     'colsample_bytree':[i/10.0 for i in range(6,10)]
+    },
+    {
+     'reg_alpha':[0, 1e-5, 1e-2, 0.1, 1, 100]
+    }
+    ]
+    for param_test in param_tests:
+        clf = tune_cls_para(clf, train, features, target='target', param_test=param_test, cv=folds, n_jobs=-1)
+
+    #%% Train the classifier with optimal parameters
+    param = {'learning_rate': 0.001,
+            'n_estimators': 50000}
+    clf.set_params(**param)
+    clf, score = xgb_fit(clf, train, features, target='target', folds=folds, early_stopping_rounds=10000)
+    #%% Model prediction
+    for df in [train, test]:
+        df.loc[:, 'OD'] = clf.predict_proba(df[features])[:, 1]
+        df.loc[:, 'antigen'] = 'combined'
+        df.loc[:, 'antigen type'] = 'Diagnostic'
+    #%% transform the dataframe back to long format for plotting
+    test_keys = test.drop(features + ['target'], axis=1)
+    plot_xgb_fscore(clf, output_dir=fig_path, output_fname='xgb_feature_importance')
+    antigen_list = ['SARS CoV2 N 50', 'SARS CoV2 RBD 250', 'SARS CoV2 spike 62.5']
+    suffix = '_'.join([pipeline, norm_antigen, 'norm_per_plate', 'mean_rands', str(rand_seed), 'low_c', 'ci'])
+    test_keys = test_keys[['plate ID', 'well_id']].drop_duplicates()
+    roc_df = df_norm.copy()
+    slice_cols = ['serum ID', 'antigen type', 'antigen']
+    slice_keys = [sera_roc_list, ['Diagnostic'], antigen_list]
+    slice_actions = ['drop', 'keep', 'keep']
+    fpr = 0.05
+    ci = 95
+    hue = 'pipeline'
+    for col, action, key in zip(slice_cols, slice_actions, slice_keys):
+        roc_df = slice_df(roc_df, action, col, key)
+    roc_df = pd.merge(test_keys, roc_df, how='left', on=['plate ID', 'well_id'])
+    #%% compute ROC curves and AUC
+    roc_df = roc_df.groupby(['antigen', 'serum ID', 'well_id', 'plate ID',
+                        'serum type', 'serum dilution', 'pipeline', 'secondary ID',
+                        'secondary dilution'])['OD'].mean()
+    roc_df = roc_df.reset_index()
+    roc_df = pd.concat([test, roc_df])
+    _ = roc_plot_grid(roc_df, fig_path, '_'.join(['ROC', suffix]), 'pdf', col_wrap=4, ci=ci, fpr=fpr, hue=hue)
```

Comment:
Do you ever need to return something from a plot function? Maybe consider dropping the return?

Reply:
Yes the function also returns a dataframe

Reply:
Yes the function also returns a dataframe